File: problems/CSES/Range_Queries/range_interval_queries/range_interval_queries.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
#define pb push_back

class FenwickTree {
private:
    int n;
    vector<ll> bit;

public:
    explicit FenwickTree(int size) : n(size), bit(size + 1, 0) {}

    explicit FenwickTree(const vector<int>& values)
        : n(static_cast<int>(values.size())), bit(n + 1, 0) {

        for (int i = 1; i <= n; i++) {
            bit[i] += values[i - 1];

            int parent = i + (i & -i);

            if (parent <= n) {
                bit[parent] += bit[i];
            }
        }
    }

    void add(int i, ll value) {
        while (i <= n) {
            bit[i] += value;
            i += i & -i;
        }
    }

    void subtract(int i, ll value) {
        add(i, -value);
    }

    ll prefix_sum(int i) {
        ll res = 0;

        while (i > 0) {
            res += bit[i];
            i -= i & -i;
        }

        return res;
    }

    ll range_sum(int left, int right) {
        return prefix_sum(right) - prefix_sum(left - 1);
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, q;
    cin >> n >> q;

    vector<int> x(n);
    for (int &v : x) cin >> v;

    vector<pair<int, int>> elements;
    for (int i = 0; i < n; ++i) elements.pb({x[i], i + 1});

    sort(elements.begin(), elements.end());

    vector<tuple<int, int, int, int, int>> events;
    for (int qi = 0; qi < q; ++qi) {
        int a, b, c, d;
        cin >> a >> b >> c >> d;

        events.pb({d, qi, 1, a, b});
        events.pb({c - 1, qi, -1, a, b});
    }

    sort(events.begin(), events.end());

    FenwickTree fw(n);

    vector<int> res(q, 0);

    int j = 0;
    for (auto [threshold, qi, sign, a, b] : events) {
        while (j < n && elements[j].first <= threshold) {
            int pos = elements[j].second;

            fw.add(pos, 1);

            j++;
        }

        res[qi] += sign * fw.range_sum(a, b);
    }

    for (int num : res) cout << num << '\n';

    return 0;
}

File: problems/CSES/Range_Queries/range_interval_queries/range_interval_queries.py
import sys
# sys.setrecursionlimit(200000)

input = sys.stdin.readline

# Russian
'''
Решение для этой задачи логически правильное, но изза того что Python не очень быстрый,
легче написать решение для этоц задачи на C++.
'''

# English
'''
The solution to this problem is logically correct, but since Python isn't very fast,
it's easier to write a solution to this problem in C++.
'''

class FenwickTree:
    def __init__(self, n):
        self.n = n
        self.bit = [0] * (n + 1)

    # Add function (also known as update),
    # and ofcourse you can use it like -value for subtract from bit, but I think it's not that cool
    def add(self, i, value):
        while i <= self.n:
            self.bit[i] += value
            i += i & -i

    def subtract(self, i, value):
        while i <= self.n:
            self.bit[i] -= value
            i += i & -i

    def prefix_sum(self, i):
        prefix_sum = 0

        while i > 0:
            prefix_sum += self.bit[i]
            i -= i & -i

        return prefix_sum

    def range_sum(self, start, end):
        return self.prefix_sum(end) - self.prefix_sum(start - 1)

def main():
    n, q = map(int, input().split())
    x = list(map(int, input().split()))

    elements = sorted((x[i], i + 1) for i in range(n))
    
    events = []
    for qi in range(q):
        a, b, c, d = map(int, input().split())

        events.append((d, qi, 1, a, b))
        events.append((c - 1, qi, -1, a, b))

    events.sort()

    fw = FenwickTree(n)
    res = [0] * q

    j = 0

    for threshold, qi, sign, a, b in events:
        while j < n and elements[j][0] <= threshold:
            _, pos = elements[j]
            fw.add(pos, 1)
            j += 1

        res[qi] += sign * fw.range_sum(a, b)

    print("\n".join(map(str, res)))

main()
